### ISSUE TITLE
No way to add elements for future generations causing fighter NFT mint functions to always revert

### ISSUE THREAD

**Issue by c4-bot-10** [code-423n4/2024-02-ai-arena-findings#630]
# Lines of code

https://github.com/code-423n4/2024-02-ai-arena/blob/cd1a0e6d1b40168657d1aaee8223dc050e15f8cc/src/FighterFarm.sol#L84-L85
https://github.com/code-423n4/2024-02-ai-arena/blob/cd1a0e6d1b40168657d1aaee8223dc050e15f8cc/src/FighterFarm.sol#L110
https://github.com/code-423n4/2024-02-ai-arena/blob/cd1a0e6d1b40168657d1aaee8223dc050e15f8cc/src/FighterFarm.sol#L470

# Vulnerability details

## Impact
`FighterFarm.sol` provides `_ownerAddress` the ability to increment fighter generations, meaning the contract is designed to handle generations > 0. However nowhere in the code can the owner or any other user actually add elements for future generations ('Elemental Powers' are a key component in each fighter NFT). Only generation 0 elements are defined as they are hard-coded in the constructor:

```solidity
    constructor(address ownerAddress, address delegatedAddress, address treasuryAddress_)
        ERC721("AI Arena Fighter", "FTR")
    {
        _ownerAddress = ownerAddress;
        _delegatedAddress = delegatedAddress;
        treasuryAddress = treasuryAddress_;
        numElements[0] = 3;
    } 
```
https://github.com/code-423n4/2024-02-ai-arena/blob/cd1a0e6d1b40168657d1aaee8223dc050e15f8cc/src/FighterFarm.sol#L104-L111

This means that as fighter type generations are incremented in `FighterFarm.sol` (as intended), the key NFT fighter minting functions will always revert. This is because both `FighterFarm::claimFighters` and `FighterFarm::redeemMintPass` use the internal `FighterFarm::_createFighterBase` function. When the elements in a generation > 0 are not defined (as they can't be), the line `uint256 element = dna % numElements[generation[fighterType]];` will always revert, as num % 0 will always give EVM divide by 0 error.

```solidity
    function _createFighterBase(
        uint256 dna, 
        uint8 fighterType
    ) 
        private 
        view 
        returns (uint256, uint256, uint256) 
    {
        uint256 element = dna % numElements[generation[fighterType]];
        uint256 weight = dna % 31 + 65;
        uint256 newDna = fighterType == 0 ? dna : uint256(fighterType);
        return (element, weight, newDna);
    }
```
https://github.com/code-423n4/2024-02-ai-arena/blob/cd1a0e6d1b40168657d1aaee8223dc050e15f8cc/src/FighterFarm.sol#L462-L474

In summary, essential fighter NFT parameters can't be defined as non-zero in a contract that is clearly intended to handle generations > 0, causing core minting functionality to always fail for generations > 0.

## Tools Used
Manual inspection

## Recommended Mitigation Steps
Add the following function to `FighterFarm.sol` to add elements for future generations:

```solidity
    /// @notice Adds a new element number to the specified generation.
    function addElements(uint8 generation, uint8 _numElements) external {
        require(msg.sender == _ownerAddress);
        numElements[generation] = _numElements;
    }
```

## Assessed type

Other
